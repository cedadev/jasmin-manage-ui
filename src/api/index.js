import { useEffect, useState } from 'react';

import {
    apiFetch,
    createContextForEndpoint,
    createContextForResource,
    useResource,
    useInstance,
    useFetchPoint
} from '../rest-resource';


// Create a context for the current user and export the hook
const CurrentUser = createContextForEndpoint("/api/me/");
const CurrentUserProvider = CurrentUser.Provider;
export const useCurrentUser = CurrentUser.hook;


// All other resources require authentication before they can be fetched
// This higher-order component wraps a provider so that it is only mounted once
// the user is successfully initialised
const withAuthentication = Provider => ({ children }) => {
    const currentUser = useCurrentUser();
    // Only wrap the children in the provider if the user is authenticated
    return currentUser.initialised ? <Provider>{children}</Provider> : children;
};


// Create a context each for the top-level, read-only resources that don't change
// often and we want to share across the component tree
const Categories = createContextForResource("/api/categories/");
const Consortia = createContextForResource("/api/consortia/");
const Resources = createContextForResource("/api/resources/");

// Wrap the providers with the higher-order component for authentication
const CategoriesProvider = withAuthentication(Categories.Provider);
const ConsortiaProvider = withAuthentication(Consortia.Provider);
const ResourcesProvider = withAuthentication(Resources.Provider);

// Export the hooks for each resource
export const useCategories = Categories.hook;
export const useConsortia = Consortia.hook;
export const useResources = Resources.hook;

// Export a single provider that provides all the top-level resources
export const Provider = ({ children }) => (
    <CurrentUserProvider>
        <CategoriesProvider>
            <ConsortiaProvider>
                <ResourcesProvider>
                    {children}
                </ResourcesProvider>
            </ConsortiaProvider>
        </CategoriesProvider>
    </CurrentUserProvider>
);


// Export a hook for loading a consortium by id
export const useConsortium = (id, options) => useInstance(`/api/consortia/${id}/`, options);

// Export hooks for accessing projects
// We don't treat the project list as a top-level shared resource because:
//   1. The list, or objects within it, changes relatively often
//   2. Projects can be loaded by id that are not in the list, e.g. when the user
//      is a consortium manager, which reduces the value in caching the list
export const useProjects = options => useResource("/api/projects/", options);
export const useProject = (id, options) => useInstance(`/api/projects/${id}/`, options);

// Export hook for loading requirement by id
export const useRequirement = (id, options) => useInstance(`/api/requirements/${id}/`, options);

// Export hook for loading service by id
export const useService = (id, options) => useInstance(`/api/services/${id}/`, options);

// Export a function for joining a project with an invite code
export const joinProject = async code => apiFetch('/api/join/', { method: 'POST', data: { code } });

// Export hook for loading category by id
export const useCategory = (id, options) => useInstance(`/api/categories/${id}/`, options);

// Export hook for loading resource by id
export const useResourceType = (id, options) => useInstance(`/api/resources/${id}/`, options);


// Export a hook for using a project event stream
export const useProjectEvents = (project, options) => {
    const { fetchPoint = true, autoFetch = true } = options || {};

    const [initialised, setInitialised] = useState(false);
    const [dirty, setDirty] = useState(autoFetch);
    const [fetching, setFetching] = useState(false);
    const [fetchError, setFetchError] = useState(null);
    const [data, setData] = useState([]);

    // Get the events URL from the project links
    const eventsUrl = project.data._links.events;

    // Build the stream object with fetchable methods
    const stream = {
        initialised,
        dirty,
        fetching,
        fetchError,
        data,
        // This function loads more events
        fetch: () => {
            // Build the URL to use
            const url = new URL(eventsUrl);
            // If there is a most recent event, use its time for the since parameter
            const event = data.find(_ => true);
            if( event ) url.searchParams.append("since", event.created_at);
            // Mark the fetch as in progress
            setFetching(true);
            // In order to prevent state updates after a component has been unmounted,
            // we need to be able to cancel fetches
            // To do this, we can use an abort controller
            const controller = new AbortController();
            // Start the fetch using the signal from the controller
            apiFetch(url, { signal: controller.signal })
                // If the promise resolves, update the state with the new data
                .then(newData => {
                    // Prepend the data to the existing data
                    setData([...newData, ...data]);
                    setInitialised(true);
                    setFetching(false);
                    setDirty(false);
                    setFetchError(null);
                })
                // If the promise rejects, update the state with the error
                .catch(error => {
                    // If the error is an abort error, it is because the component that
                    // owns the data has been unmounted so don't update the state
                    if( error.name === "AbortError" ) return;
                    setFetching(false);
                    setDirty(false);
                    setFetchError(error);
                });
            // Return the abort function to allow the fetch to be cancelled from outside
            return () => controller.abort();
        },
        // This function marks the stream as dirty, causing more data to be fetched
        markDirty: () => setDirty(true),
        // This function resets the stream, clearing all the data and triggering a re-fetch
        reset: () => {
            setInitialised(false);
            setDirty(true);
            setFetching(false);
            setFetchError(null);
            setData([]);
        }
    };

    // Hooks cannot be conditional, so we always have to call the fetch point hook
    // However we can effectively disable it by making sure it never sees a dirty stream
    useFetchPoint({ ...stream, dirty: fetchPoint && stream.dirty });

    // We need to be a fetch point for the event stream
    return stream;
};
