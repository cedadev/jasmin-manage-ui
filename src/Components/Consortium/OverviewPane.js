import React, { useState } from 'react';

import Card from 'react-bootstrap/Card';
import Col from 'react-bootstrap/Col';
import ProgressBar from 'react-bootstrap/ProgressBar';
import Row from 'react-bootstrap/Row';
import Table from 'react-bootstrap/Table';

import classNames from 'classnames';

import { Status, useNestedResource } from '../../rest-resource';

import { useResources } from '../../api';

import { SpinnerWithText, formatAmount, sortByKey } from '../utils';

import '../../css/quota-card.css';


const QuotaCard = ({ quota, resource }) => {
    // Store some state to determine which progress bar/table row should be highlighted
    // Can be null, "approved", "awaiting" or "provisioned"
    const [highlightedAmount, setHighlightedAmount] = useState(null);
    const { name: resourceName, units } = resource.data;
    const {
        amount,
        total_approved: totalApproved,
        total_awaiting_provisioning: totalAwaiting,
        total_provisioned: totalProvisioned
    } = quota.data;
    const totalCommitted = totalApproved + totalAwaiting + totalProvisioned;
    // If the total committed is more than the quota amount, we will show that
    // by placing a line across the bar where the quota amount is
    const progressBarMax = Math.max(amount, totalCommitted);
    // Calculate the allocated percentage
    const committedPercentage = Math.round((totalCommitted / amount) * 100);
    return (
        <Card className="mb-3 quota-card" style={{ borderWidth: '3px' }}>
            <Card.Header>
                <h5 className="mb-0">{resourceName}</h5>
            </Card.Header>
            <Card.Body>
                <Card.Title className="text-center">
                    {formatAmount(totalCommitted, units)}{" "}
                    of{" "}
                    {formatAmount(amount, units)}{" "}
                    ({committedPercentage}%)
                </Card.Title>
                <ProgressBar className="position-relative">
                    <ProgressBar
                        variant="info"
                        max={progressBarMax}
                        now={totalProvisioned}
                        className={classNames({
                            'inactive': highlightedAmount,
                            'active': highlightedAmount === "provisioned"
                        })}
                        onMouseEnter={() => setHighlightedAmount("provisioned")}
                        onMouseLeave={() => setHighlightedAmount(null)}
                    />
                    <ProgressBar
                        variant="warning"
                        max={progressBarMax}
                        now={totalAwaiting}
                        className={classNames({
                            'inactive': highlightedAmount,
                            'active': highlightedAmount === "awaiting"
                        })}
                        onMouseEnter={() => setHighlightedAmount("awaiting")}
                        onMouseLeave={() => setHighlightedAmount(null)}
                    />
                    <ProgressBar
                        variant="success"
                        max={progressBarMax}
                        now={totalApproved}
                        className={classNames({
                            'inactive': highlightedAmount,
                            'active': highlightedAmount === "approved"
                        })}
                        onMouseEnter={() => setHighlightedAmount("approved")}
                        onMouseLeave={() => setHighlightedAmount(null)}
                    />
                    {progressBarMax > amount && (
                        <div
                            style={{
                                position: 'absolute',
                                width: '3px',
                                height: '100%',
                                backgroundColor: 'red',
                                zIndex: 1000,
                                top: 0,
                                left: `${(amount / progressBarMax) * 100}%`
                            }}
                        />
                    )}
                </ProgressBar>
            </Card.Body>
            <Table>
                <tbody>
                    <tr>
                        <th>Quota amount</th>
                        <td>{formatAmount(amount, units)}</td>
                    </tr>
                    <tr
                        className={classNames({
                            'table-active': highlightedAmount === "provisioned"
                        })}
                        onMouseEnter={() => setHighlightedAmount("provisioned")}
                        onMouseLeave={() => setHighlightedAmount(null)}
                    >
                        <th>Total Provisioned</th>
                        <td>{formatAmount(totalProvisioned, units)}</td>
                    </tr>
                    <tr
                        className={classNames({
                            'table-active': highlightedAmount === "awaiting"
                        })}
                        onMouseEnter={() => setHighlightedAmount("awaiting")}
                        onMouseLeave={() => setHighlightedAmount(null)}
                    >
                        <th>Total Awaiting Provisioning</th>
                        <td>{formatAmount(totalAwaiting, units)}</td>
                    </tr>
                    <tr
                        className={classNames({
                            'table-active': highlightedAmount === "approved"
                        })}
                        onMouseEnter={() => setHighlightedAmount("approved")}
                        onMouseLeave={() => setHighlightedAmount(null)}
                    >
                        <th>Total Approved</th>
                        <td>{formatAmount(totalApproved, units)}</td>
                    </tr>
                </tbody>
            </Table>
        </Card>
    );
};


const OverviewPane = ({ consortium }) => {
    // To render the overview, we need the quotas and the resources
    const quotas = useNestedResource(consortium, "quotas");
    const resources = useResources();
    return (
        <Status.Many fetchables={[quotas, resources]}>
            <Status.Loading>
                <div className="d-flex justify-content-center my-5">
                    <SpinnerWithText>Loading quotas...</SpinnerWithText>
                </div>
            </Status.Loading>
            <Status.Unavailable>
                <Status.Throw />
            </Status.Unavailable>
            <Status.Available>
                {([quotaData, resourceData]) => {
                    // Sort the quotas by resource name
                    const sortedQuotas = sortByKey(
                        Object.values(quotaData),
                        quota => resourceData[quota.data.resource].data.name
                    );
                    return (
                        <Row xs={1} lg={2}>
                            {sortedQuotas.map(quota => (
                                <Col key={quota.data.id}>
                                    <QuotaCard
                                        quota={quota}
                                        resource={resourceData[quota.data.resource]}
                                    />
                                </Col>
                            ))}
                        </Row>
                    );
                }}
            </Status.Available>
        </Status.Many>
    );
};


export default OverviewPane;
