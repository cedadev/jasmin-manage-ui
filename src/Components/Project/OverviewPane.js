import React, { useState } from 'react';

import { useNavigate } from 'react-router-dom';
import Alert from 'react-bootstrap/Alert';
import Card from 'react-bootstrap/Card';
import Button from 'react-bootstrap/Button';
import Modal from 'react-bootstrap/Modal';
import Form from 'react-bootstrap/Form';
import { useNotifications } from 'react-bootstrap-notify';

import classNames from 'classnames';

import moment from 'moment';

import ReactMarkdown from 'react-markdown';


import { Form as ResourceForm } from '../../rest-resource';

import { notificationFromError, MarkdownEditor } from '../utils';

import {
    Status,
    useNestedResource,
    useAggregateResource,
    useFetchPoint,
    useEnsureInitialised
} from '../../rest-resource';

import { useResources } from '../../api';

import { formatAmount, sortByKey } from '../utils';

import { useProjectPermissions } from './actions';

import '../../css/project-detail.css';


const TimelineItem = ({ children }) => (
    <div className="timeline-item">
        <div className="timeline-item-container">
            <div className="timeline-item-content">
                {children}
            </div>
        </div>
    </div>
);


const ProjectDescription = ({ project }) => {

    const notify = useNotifications();
    const [modalVisible, setModalVisible] = useState(false);
    const showModal = () => setModalVisible(true);
    const hideModal = () => setModalVisible(false);
    // Get the project permissions for description edit
    const {
        canEditRequirements,
        canSubmitForReview,
        canRequestChanges,
        canSubmitForProvisioning
    } = useProjectPermissions(project);

    const navigate = useNavigate();
    const handleError = error => {
        notify(notificationFromError(error));
        hideModal();
    };
    // Mark the markdown editor control into a resource form control
    const MarkdownEditorControl = ResourceForm.Controls.asControl(
        MarkdownEditor,
        evt => evt.target.value,
    );
    return(<>
    <Card>
        <Card.Header><strong>Project description</strong></Card.Header>
        <Card.Body className="markdown">
            <ReactMarkdown children={project.data.description} />
            {canSubmitForReview && (<>
                <Button onClick={showModal} size="sm" variant="outline-primary" style={{float: 'right'}}>edit</Button>
                
                <Modal show={modalVisible} backdrop="static" keyboard={false}>
                    <ResourceForm.UpdateInstanceForm
                        instance={project}
                        field={project.data.description}
                        onSuccess={hideModal}
                        onError={handleError}
                        onCancel={hideModal}
                    >
                        <Modal.Header>
                            <Modal.Title>Edit project description</Modal.Title>
                        </Modal.Header>
                        <Modal.Body>
                            
                            <Form.Group controlId="description">
                                <Form.Label>Description</Form.Label>
                                <Form.Control as={MarkdownEditorControl} required />

                                <ResourceForm.Controls.ErrorList />
                            </Form.Group>
                        </Modal.Body>
                        <Modal.Footer>
                            <ResourceForm.Controls.CancelButton>Cancel</ResourceForm.Controls.CancelButton>
                            <ResourceForm.Controls.SubmitButton>Confirm</ResourceForm.Controls.SubmitButton>
                        </Modal.Footer>
                    </ResourceForm.UpdateInstanceForm>
                </Modal>
                </>
            )}
        </Card.Body>
    </Card>
    </>
    )
};


const Comment = ({ project, item }) => {
    const displayName = item.data.user.last_name ?
        `${item.data.user.first_name} ${item.data.user.last_name}` :
        item.data.user.username;
    const createdAt = moment(item.data.created_at).format('D MMMM');
    return (
        <Card className="comment">
            <Card.Header>
                <strong>{displayName}</strong> <em>commented on {createdAt}</em>
            </Card.Header>
            <Card.Body className="markdown">
                <ReactMarkdown children={item.data.content} />
            </Card.Body>
        </Card>
    );
};


const EventIcon = ({ icon, variant }) => {
    const variantClassName = !!variant ? `text-${variant}` : undefined;
    return (
        <span className={classNames("event-icon", "fa-stack", variantClassName)}>
            <i className="fas fa-circle fa-stack-2x" />
            <i className={classNames("fas", "fa-stack-1x", "fa-inverse", icon)} />
        </span>
    );
};


const EventText = props => <span className="event-text" {...props} />;


const ProjectEvent = ({ children, item, className }) => {
    const createdBy = item.user.last_name ?
        `${item.user.first_name} ${item.user.last_name}` :
        item.user.username;
    const createdAt = moment(item.created_at).format('D MMMM');
    return (
        <div className={classNames("event", className)}>
            {children(createdBy, createdAt)}
        </div>
    );
};


const ResourceEvent = ({ children, project, requirements, item, ...props }) => {
    const resources = useResources();
    // Force the requirements to load
    useEnsureInitialised(requirements);
    return (
        <Status.Many fetchables={[resources, requirements]}>
            <Status.Loading>
                <div className="event text-muted">
                    <EventIcon icon="fa-sync-alt fa-spin" />
                    <EventText>Loading requirements...</EventText>
                </div>
            </Status.Loading>
            <Status.Unavailable>
                <div className="event text-danger font-weight-bold">
                    <EventIcon icon="fa-bomb" />
                    <EventText>Unable to load requirements</EventText>
                </div>
            </Status.Unavailable>
            <Status.Available>
                {([resourceData, requirementData]) => {
                    if (requirementData[item.target_id]) {
                        const requirement = requirementData[item.target_id];
                        const resource = resourceData[requirement.data.resource];
                        const amount = formatAmount(requirement.data.amount, resource.data.units);
                        return (
                            <ProjectEvent item={item} {...props}>
                                {(createdBy, createdAt) => children(
                                    amount,
                                    resource.data.name,
                                    createdBy,
                                    createdAt
                                )}
                            </ProjectEvent>
                        );
                    };
                    return (
                        <div className="event">
                            <EventIcon icon="fa-question" variant="light" />
                            <EventText className="text-muted">
                                Event <strong>missing</strong> on {moment(item.created_at).format('D MMMM')}
                            </EventText>
                        </div>
                    );
                }}
            </Status.Available>
        </Status.Many>
    );
};


// Map of timeline events to a component
const TimelineEvents = {
    'jasmin_manage.project.created': props => (
        <ProjectEvent {...props}>
            {(createdBy, createdAt) => (<>
                <EventIcon icon="fa-plus" variant="info" />
                <EventText>
                    Project <strong>created</strong>{" "}
                    on {createdAt}{" "}
                    by <strong>{createdBy}</strong>
                </EventText>
            </>)}
        </ProjectEvent>
    ),
    'jasmin_manage.project.submitted_for_review': props => (
        <ProjectEvent {...props}>
            {(createdBy, createdAt) => (<>
                <EventIcon icon="fa-question" variant="primary" />
                <EventText>
                    Project <strong>submitted for review</strong>{" "}
                    on {createdAt}{" "}
                    by <strong>{createdBy}</strong>
                </EventText>
            </>)}
        </ProjectEvent>
    ),
    'jasmin_manage.project.changes_requested': props => (
        <ProjectEvent {...props}>
            {(createdBy, createdAt) => (<>
                <EventIcon icon="fa-exclamation" variant="danger" />
                <EventText>
                    <strong>Changes requested</strong>{" "}
                    on {createdAt}{" "}
                    by <strong>{createdBy}</strong>
                </EventText>
            </>)}
        </ProjectEvent>
    ),
    'jasmin_manage.project.submitted_for_provisioning': props => (
        <ProjectEvent {...props}>
            {(createdBy, createdAt) => (<>
                <EventIcon icon="fa-check" variant="success" />
                <EventText>
                    Project <strong>submitted for provisioning</strong>{" "}
                    on {createdAt}{" "}
                    by <strong>{createdBy}</strong>
                </EventText>
            </>)}
        </ProjectEvent>
    ),
    'jasmin_manage.requirement.provisioned': props => (
        <ResourceEvent {...props}>
             {(amount, resourceName, createdBy, createdAt) => (<>
                <EventIcon icon="fa-layer-group" variant="secondary" />
                <EventText>
                    <strong>{amount} {resourceName} provisioned</strong>{" "}
                    on {createdAt}
                </EventText>
            </>)}
        </ResourceEvent>
    ),
    'jasmin_manage.requirement.decommissioned': props => (
        <ResourceEvent {...props} className="text-muted">
             {(amount, resourceName, createdBy, createdAt) => (<>
                <EventIcon icon="fa-power-off" />
                <EventText>
                    <strong>{amount} {resourceName} decommissioned</strong>{" "}
                    on {createdAt}
                </EventText>
            </>)}
        </ResourceEvent>
    ),
};


const getTimelineData = (commentData, eventData) => {
    const timelineData = [
        // Start with the comment data
        ...Object.values(commentData).map(comment => ({
            id: `comment-${comment.data.id}`,
            data: comment,
            component: Comment,
            createdAt: comment.data.created_at
        })),
        // Add the event data, but only those with a renderer
        ...eventData
            .filter(event => TimelineEvents.hasOwnProperty(event.event_type))
            .map(event => ({
                id: `event-${event.id}`,
                data: event,
                component: TimelineEvents[event.event_type],
                createdAt: event.created_at
            }))
    ];
    // Return the data sorted by createdAt
    return sortByKey(timelineData, item => item.createdAt, true);
};


const ProjectTimeline = ({ project, events }) => {
    // Load the comments
    const comments = useNestedResource(project, "comments");

    // This is the fetch point for the events
    useFetchPoint(events);

    // In order to render some events, we need the project services and requirements
    // We want to fetch them once for all events, so this needs to be the fetch point
    // However we don't want to load them if we don't need them, so don't auto-fetch
    // This means that individual event components must force them to load using
    // "useEnsureInitialised" if they need them
    const services = useNestedResource(project, "services", { autoFetch: false });
    const requirements = useAggregateResource(services, "requirements", { autoFetch: false });

    return (<>
        {/* If either the comments or events are loading, include an item at the top */}
        {(comments.fetching || events.fetching ) && (
            <TimelineItem>
                <div className="event text-muted">
                    <EventIcon icon="fa-sync-alt fa-spin" />
                    <EventText>Loading timeline items...</EventText>
                </div>
            </TimelineItem>
        )}
        <Status.Many fetchables={[comments, events]}>
            {/* No loading section as this is covered by the item above */}
            <Status.Unavailable>
                <TimelineItem>
                    <div className="event text-danger font-weight-bold">
                        <EventIcon icon="fa-bomb" />
                        <EventText>Error loading timeline data</EventText>
                    </div>
                </TimelineItem>
            </Status.Unavailable>
            <Status.Available>
                {([commentData, eventData]) => (
                    // Render each timeline item with the specified component
                    getTimelineData(commentData, eventData).map(item => (
                        <TimelineItem key={item.id}>
                            <item.component
                                project={project}
                                services={services}
                                requirements={requirements}
                                item={item.data}
                            />
                        </TimelineItem>
                    ))
                )}
            </Status.Available>
        </Status.Many>
    </>);
};


export const OverviewPane = ({ project, events }) => (
    <div className="timeline">
        <TimelineItem>
            <ProjectDescription project={project} />
        </TimelineItem>
        <ProjectTimeline project={project} events={events} />
    </div>
);
