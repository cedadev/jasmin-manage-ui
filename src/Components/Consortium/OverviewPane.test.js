import React from "react";
import { render, screen, waitFor, within } from "@testing-library/react";
import renderer from "react-test-renderer";
import { BrowserRouter as Router } from "react-router-dom";

import OverviewPane from "./OverviewPane";
import { useResources } from "../../api";

// Mock the API hooks
jest.mock("../../api");

// Helper function to render the component
const renderOverviewPane = (quotas) => {
  return render(
    <Router>
      <OverviewPane quotas={quotas} />
    </Router>
  );
};

describe("OverviewPane Component", () => {
  // Mock data for resources
  const mockResources = {
    data: {
      1: {
        data: {
          id: 1,
          name: "Resource A",
          units: "GB",
        },
      },
      2: {
        data: {
          id: 2,
          name: "Resource B",
          units: "TB",
        },
      },
    },
  };

  // Mock data for quotas
  const mockQuotas = {
    data: {
      1: {
        data: {
          id: 1,
          resource: 1,
          amount: 100,
          total_approved: 50,
          total_awaiting_provisioning: 20,
          total_provisioned: 30,
        },
      },
      2: {
        data: {
          id: 2,
          resource: 2,
          amount: 200,
          total_approved: 150,
          total_awaiting_provisioning: 30,
          total_provisioned: 20,
        },
      },
    },
  };

  beforeEach(() => {
    useResources.mockReturnValue(mockResources);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("Should match the snapshot", () => {
    const component = renderer.create(
      <Router>
        <OverviewPane quotas={mockQuotas} />
      </Router>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Should display loading state when fetching", async () => {
    useResources.mockReturnValue({
      data: null,
      fetching: true,
      initialised: false,
    });

    renderOverviewPane(mockQuotas);
    expect(screen.getByText("Loading quotas...")).toBeInTheDocument();
  });

  it("Should display quota cards with correct information", async () => {
    // Mock resources with initialized state
    useResources.mockReturnValue({
      data: mockResources.data,
      fetching: false,
      fetchError: null,
      initialised: true,
    });

    // Mock quotas with initialized state
    const quotas = {
      data: mockQuotas.data,
      fetching: false,
      fetchError: null,
      initialised: true,
    };

    renderOverviewPane(quotas);

    // Wait for the component to render and verify content
    await waitFor(() => {
      // Check for resource names
      expect(screen.getByText("Resource A")).toBeInTheDocument();
    });

    // Now check the rest of the quota information
    const quotaCardA = screen.getByText("Resource A").closest(".card");
    const quotaCardB = screen.getByText("Resource B").closest(".card");

    // Check quota card A content
    within(quotaCardA).getByText("100 GB"); // Quota amount
    within(quotaCardA).getByText("30 GB"); // Total Provisioned
    within(quotaCardA).getByText("20 GB"); // Total Awaiting Provisioning
    within(quotaCardA).getByText("50 GB"); // Total Approved

    // Check quota card B content
    within(quotaCardB).getByText("200 TB"); // Quota amount
    within(quotaCardB).getByText("20 TB"); // Total Provisioned
    within(quotaCardB).getByText("30 TB"); // Total Awaiting Provisioning
    within(quotaCardB).getByText("150 TB"); // Total Approved
  });

  it("Should display warning border for quotas between 70% and 90% usage", async () => {
    useResources.mockReturnValue({
      data: {
        1: {
          data: {
            id: 1,
            name: "Resource A",
            units: "GB",
          },
        },
      },
      fetching: false,
      fetchError: null,
      initialised: true,
    });

    const warningQuotas = {
      data: {
        1: {
          data: {
            id: 1,
            resource: 1,
            amount: 100,
            total_approved: 60,
            total_awaiting_provisioning: 10,
            total_provisioned: 5,
          },
        },
      },
      fetching: false,
      fetchError: null,
      initialised: true,
    };

    renderOverviewPane(warningQuotas);

    // Wait for loading to complete
    await waitFor(() => {
      expect(screen.queryByText("Loading quotas...")).not.toBeInTheDocument();
    });

    // Find card by heading and verify border class
    const quotaCard = screen
      .getByRole("heading", { name: "Resource A" })
      .closest(".card");
    expect(quotaCard).toHaveClass("quota-card", "border-warning");

    // Verify percentage is correct
    expect(screen.getByText(/75%/)).toBeInTheDocument();
  });

  it("Should display danger border for quotas above 90% usage", async () => {
    // Mock resources with the required resource for the test
    useResources.mockReturnValue({
      data: {
        1: {
          data: {
            id: 1,
            name: "Resource A",
            units: "GB",
          },
        },
      },
      initialised: true,
      fetching: false,
    });

    const dangerQuotas = {
      data: {
        1: {
          data: {
            id: 1,
            resource: 1,
            amount: 100,
            total_approved: 80,
            total_awaiting_provisioning: 10,
            total_provisioned: 5,
          },
        },
      },
      initialised: true,
      fetching: false,
    };

    renderOverviewPane(dangerQuotas);

    // Wait for the component to render and check for quota card
    await waitFor(() => {
      expect(screen.getByText("Resource A")).toBeInTheDocument();
    });

    const quotaCard = screen.getByText("Resource A").closest(".quota-card");
    expect(quotaCard).toHaveClass("border-danger");
  });
});
