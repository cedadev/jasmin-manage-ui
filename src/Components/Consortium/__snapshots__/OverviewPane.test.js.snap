// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OverviewPane Component Should match the snapshot 1`] = `
<div
  className="d-flex justify-content-center my-5"
>
  <div
    style={
      {
        "fontSize": "120%",
      }
    }
  >
    <i
      className="fas fa-spin fa-sync-alt mr-2 fa-lg"
    />
    Loading quotas...
  </div>
</div>
`;
